File: src/Fortran/H5Part_io.c
/*
  Copyright (c) 2006-2015, The Regents of the University of California,
  through Lawrence Berkeley National Laboratory (subject to receipt of any
  required approvals from the U.S. Dept. of Energy) and the Paul Scherrer
  Institut (Switzerland).  All rights reserved.

  License: see file COPYING in top level of source distribution.
*/

#include "h5_private.h"
#include "h5core/h5_log.h"

#include "h5core/h5u_io.h"

/*==================Writing data ============*/
#define h5pt_writedata_r8 FC_MANGLING (					\
                h5pt_writedata_r8,                                      \
                H5PT_WRITEDATA_R8 )
h5_int64_t
h5pt_writedata_r8 (
	const h5_int64_t* const fh,
	const char* const name,
	const h5_float64_t* const data,
	const int l_name
        ) {
	h5_file_t f = h5_filehandlefor2c (fh);
	H5_API_ENTER (h5_int64_t,
                      "fh=%p, name='%s', data=%p, l_name=%d",
                      (h5_file_p)f, name, data, l_name);
	char *name2 = h5_strdupfor2c ( name, l_name );
	h5_int64_t herr = h5u_write_dataset (
		f, name2, (void*)data, H5_FLOAT64_T );
	free ( name2 );
	H5_API_RETURN(herr);
}

#define h5pt_writedata_r4 FC_MANGLING (					\
                h5pt_writedata_r4,                                      \
                H5PT_WRITEDATA_R4 )
h5_int64_t
h5pt_writedata_r4 (
	const h5_int64_t* const fh,
	const char* const name,
	const h5_float32_t* const data,
	const int l_name
        ) {
	h5_file_t f = h5_filehandlefor2c (fh);
	H5_API_ENTER (h5_int64_t,
                      "fh=%p, name='%s', data=%p, l_name=%d",
                      (h5_file_p)f, name, data, l_name);
	char *name2 = h5_strdupfor2c ( name, l_name );
	h5_int64_t herr = h5u_write_dataset (
		f, name2, (void*)data, H5_FLOAT32_T );
	free ( name2 );
	H5_API_RETURN(herr);
}

#define h5pt_writedata_i8 FC_MANGLING (					\
                h5pt_writedata_i8,                                      \
                H5PT_WRITEDATA_I8 )
h5_int64_t
h5pt_writedata_i8 (
	const h5_int64_t* const fh,
	const char* const name,
	const h5_int64_t* const data,
	const int l_name
        ) {
	h5_file_t f = h5_filehandlefor2c (fh);
	H5_API_ENTER (h5_int64_t,
                      "fh=%p, name='%s', data=%p, l_name=%d",
                      (h5_file_p)f, name, data, l_name);
	char *name2 = h5_strdupfor2c ( name, l_name );
	h5_int64_t herr = h5u_write_dataset (
		f, name2, (void*)data, H5_INT64_T );
	free ( name2 );
	H5_API_RETURN(herr);
}

#define h5pt_writedata_i4 FC_MANGLING (					\
                h5pt_writedata_i4,                                      \
                H5PT_WRITEDATA_I4 )
h5_int64_t
h5pt_writedata_i4 (
	const h5_int64_t* const fh,
	const char* const name,
	const h5_int32_t* const data,
	const int l_name
        ) {
	h5_file_t f = h5_filehandlefor2c (fh);
	H5_API_ENTER (h5_int64_t,
                      "fh=%p, name='%s', data=%p, l_name=%d",
                      (h5_file_p)f, name, data, l_name);
	char *name2 = h5_strdupfor2c ( name, l_name );
	h5_int64_t herr = h5u_write_dataset (
		f, name2, (void*)data, H5_INT32_T );
	free ( name2 );
	H5_API_RETURN(herr);
}


/*==================Reading data ============*/
#define h5pt_readdata_r8 FC_MANGLING (					\
                h5pt_readdata_r8,                                       \
                H5PT_READDATA_R8 )
h5_int64_t
h5pt_readdata_r8 (
	const h5_int64_t* const fh,
	const char* const name,
	h5_float64_t* const data,
	const int l_name
	) {
	h5_file_t f = h5_filehandlefor2c (fh);
	H5_API_ENTER (h5_int64_t,
                      "fh=%p, name='%s', data=%p, l_name=%d",
                      (h5_file_p)f, name, data, l_name);
	char *name2 = h5_strdupfor2c ( name, l_name );
	h5_int64_t herr = h5u_read_dataset (
		f, name2, data, H5_FLOAT64_T );
	free ( name2 );
	H5_API_RETURN(herr);
}

#define h5pt_readdata_r4 FC_MANGLING (					\
                h5pt_readdata_r4,                                       \
                H5PT_READDATA_R4 )
h5_int64_t
h5pt_readdata_r4 (
	const h5_int64_t* const fh,
	const char* const name,
	h5_float32_t* const data,
	const int l_name
	) {
	h5_file_t f = h5_filehandlefor2c (fh);
	H5_API_ENTER (h5_int64_t, 
                      "fh=%p, name='%s', data=%p, l_name=%d",
                      (h5_file_p)f, name, data, l_name);
	char *name2 = h5_strdupfor2c ( name, l_name );
	h5_int64_t herr = h5u_read_dataset (
		f, name2, data, H5_FLOAT32_T );
	free ( name2 );
	H5_API_RETURN(herr);
}

#define h5pt_readdata_i8 FC_MANGLING (					\
                h5pt_readdata_i8,                                       \
                H5PT_READDATA_I8 )
h5_int64_t
h5pt_readdata_i8 (
	const h5_int64_t* const fh,
	const char* const name,
	h5_int64_t* const data,
	const int l_name
	) {
	h5_file_t f = h5_filehandlefor2c (fh);
	H5_API_ENTER (h5_int64_t,
                      "fh=%p, name='%s', data=%p, l_name=%d",
                      (h5_file_p)f, name, data, l_name);
	char *name2 = h5_strdupfor2c ( name, l_name );
	h5_int64_t herr = h5u_read_dataset (
		f, name2, data, H5_INT64_T );

	free ( name2 );
	H5_API_RETURN(herr);
}

#define h5pt_readdata_i4 FC_MANGLING (					\
                h5pt_readdata_i4,                                       \
                H5PT_READDATA_I4 )
h5_int64_t
h5pt_readdata_i4 (
	const h5_int64_t* const fh,
	const char* const name,
	h5_int32_t* const data,
	const int l_name
	) {
	h5_file_t f = h5_filehandlefor2c (fh);
	H5_API_ENTER (h5_int64_t,
                      "hf=%p, name='%s', data=%p, l_name=%d",
                      (h5_file_p)f, name, data, l_name);
	char *name2 = h5_strdupfor2c ( name, l_name );
	h5_int64_t herr = h5u_read_dataset (
		f, name2, data, H5_INT32_T );
	free ( name2 );
	H5_API_RETURN(herr);
}

File: h5hut/src/include/H5Part_io.f90
!
!  Copyright (c) 2006-2016, The Regents of the University of California,
!  through Lawrence Berkeley National Laboratory (subject to receipt of any
!  required approvals from the U.S. Dept. of Energy) and the Paul Scherrer
!  Institut (Switzerland).  All rights reserved.!
!
!  License: see file COPYING in top level of source distribution.
!
  INTERFACE
     !> \ingroup h5part_f90_api
     !! \addtogroup h5part_io_f
     !! @{

     !>
     !! See \ref H5PartWriteDataFloat64
     !! \return 0 on success or error code

     INTEGER*8 FUNCTION h5pt_writedata_r8 ( filehandle, name, data )
       INTEGER*8, INTENT(IN) :: filehandle         !< the handle returned during file open
       CHARACTER(LEN=*), INTENT(IN) :: name        !< the name of the dataset
       REAL*8, INTENT(IN) :: data(*)               !< the array of float64 data to write
     END FUNCTION h5pt_writedata_r8

     !>
     !! See \ref H5PartWriteDataFloat32
     !! \return 0 on success or error code

     INTEGER*8 FUNCTION h5pt_writedata_r4 ( filehandle, name, data )
       INTEGER*8, INTENT(IN) :: filehandle         !< the handle returned during file open
       CHARACTER(LEN=*), INTENT(IN) :: name        !< the name of the dataset
       REAL, INTENT(IN) :: data(*)                 !< the array of float32 data to write
     END FUNCTION h5pt_writedata_r4

     !>
     !! See \ref H5PartWriteDataInt64
     !! \return 0 on success or error code

     INTEGER*8 FUNCTION h5pt_writedata_i8 ( filehandle, name, data )
       INTEGER*8, INTENT(IN) :: filehandle         !< the handle returned during file open
       CHARACTER(LEN=*), INTENT(IN) :: name        !< the name of the dataset
       INTEGER*8, INTENT(IN) :: data(*)            !< the array of int64 data to write
     END FUNCTION h5pt_writedata_i8

     !>
     !! See \ref H5PartWriteDataInt32
     !! \return 0 on success or error code

     INTEGER*8 FUNCTION h5pt_writedata_i4 ( filehandle, name, data )
       INTEGER*8, INTENT(IN) :: filehandle         !< the handle returned during file open
       CHARACTER(LEN=*), INTENT(IN) :: name        !< the name of the dataset
       INTEGER, INTENT(IN) :: data(*)              !< the array of int32 data to write
     END FUNCTION h5pt_writedata_i4


     !>
     !! See \ref H5PartReadDataFloat64
     !! \return 0 on success or error code

     INTEGER*8 FUNCTION h5pt_readdata_r8 (filehandle,name,data)
       INTEGER*8, INTENT(IN) :: filehandle         !< the handle returned during file open
       CHARACTER(LEN=*), INTENT(IN) :: name        !< the name of the dataset
       REAL*8, INTENT(OUT) :: data(*)              !< array to read float64 data into
     END FUNCTION h5pt_readdata_r8

     !>
     !! See \ref H5PartReadDataFloat32
     !! \return 0 on success or error code

     INTEGER*8 FUNCTION h5pt_readdata_r4 (filehandle,name,data)
       INTEGER*8, INTENT(IN) :: filehandle         !< the handle returned during file open
       CHARACTER(LEN=*), INTENT(IN) :: name        !< the name of the dataset
       REAL, INTENT(OUT) :: data(*)                !< array to read float32 data into
     END FUNCTION h5pt_readdata_r4

     !>
     !! See \ref H5PartReadDataInt64
     !! \return 0 on success or error code

     INTEGER*8 FUNCTION h5pt_readdata_i8 (filehandle,name,data)
       INTEGER*8, INTENT(IN) :: filehandle         !< the handle returned during file open
       CHARACTER(LEN=*), INTENT(IN) :: name        !< the name of the dataset
       INTEGER*8, INTENT(OUT) :: data(*)           !< array to read int64 data into
     END FUNCTION h5pt_readdata_i8

     !>
     !! See \ref H5PartReadDataInt32
     !! \return 0 on success or error code

     INTEGER*8 FUNCTION h5pt_readdata_i4 (filehandle,name,data)
       INTEGER*8, INTENT(IN) :: filehandle         !< the handle returned during file open
       CHARACTER(LEN=*), INTENT(IN) :: name        !< the name of the dataset
       INTEGER, INTENT(OUT) :: data(*)             !< array to read int32 data into
     END FUNCTION h5pt_readdata_i4

     !> @}
     
  END INTERFACE
